gpflow/kullback_leiblers.py

```diff
@@ -80,26 +83,49 @@ def gauss_kl(q_mu, q_sqrt, K=None):
         trace = tf.reduce_sum(tf.square(Lq))
     else:
         if diag:
-            M = tf.shape(Lp)[0]
-            Lp_inv = tf.matrix_triangular_solve(
-                Lp, tf.eye(M, dtype=settings.float_type), lower=True)
-            K_inv = tf.matrix_triangular_solve(
-                tf.transpose(Lp), Lp_inv, lower=False)
-            trace = tf.reduce_sum(
-                tf.expand_dims(tf.matrix_diag_part(K_inv), 1) * tf.square(q_sqrt))
+
+            if K.get_shape().ndims == 3:
+                LpT = tf.transpose(Lp,(0,2,1)) # L x M x M
+                Lp_inv = tf.matrix_triangular_solve(Lp,
+                        tf.matrix_diag(tf.ones((L, M), dtype=settings.float_type)), # L x M x M
+                        lower=True) # L x M x M
+                K_inv = tf.transpose(
+                    tf.matrix_diag_part(
+                        tf.matrix_triangular_solve(LpT, Lp_inv, lower=False) # L x M x M
+                    ) # L x M
+                ) # M x L
+            else:
+                LpT = tf.transpose(Lp) # M x M
+                Lp_inv = tf.matrix_triangular_solve(Lp,
+                        tf.eye(M, dtype=settings.float_type), #  M x M
+                        lower=True) # M x M
+                K_inv = tf.matrix_diag_part(
+                    tf.matrix_triangular_solve(LpT, Lp_inv, lower=False)  # M x M
+                )[:,None] # M x 1 (implicit extension to broadcast)
```

Comment:
would be good to minimise code divergence and only have a single call to matrix_triangular_solve and matrix_diag_part, and only set arguments to these functions in the if: else: